checker/src/org/checkerframework/checker/lock/wholefiles/LockVisitor.java

```diff
@@ -0,0 +1,580 @@
+
+            if (treeElement != null && treeElement.getKind() == ElementKind.METHOD) { // Method calls are not dereferences.
+                isMethodCall = true;
+            }
+        }
+
+        if (!isMethodCall) {
+            ExpressionTree expr = null;
+
+            switch(treeKind) {
+                case ARRAY_ACCESS:
+                    expr = ((ArrayAccessTree) tree).getExpression();
+                    break;
+                case MEMBER_SELECT:
+                    expr = ((MemberSelectTree) tree).getExpression();
+                    break;
+                default:
+                    expr = tree;
+                    break;
+            }
+
+            Element invokedElement = TreeUtils.elementFromUse(expr);
+
+            AnnotatedTypeMirror receiverAtm = atypeFactory.getReceiverType(tree);
+            Node node = atypeFactory.getNodeForTree(tree);
+
+            // Is the receiver of the expression being accessed the same as the receiver of the enclosing method?
+            boolean receiverIsThatOfEnclosingMethod = false;
+
+            if (node instanceof FieldAccessNode) {
+                Node receiverNode = ((FieldAccessNode) node).getReceiver();
+                if (receiverNode instanceof ExplicitThisLiteralNode ||
+                    receiverNode instanceof ImplicitThisLiteralNode ||
+                    receiverNode instanceof ThisLiteralNode) {
+                    receiverIsThatOfEnclosingMethod = true;
+                }
+            }
+
+            if (expr != null && invokedElement != null && receiverAtm != null) {
+                boolean skipCheckPreconditions = false;
+
+                AnnotationMirror gb = receiverAtm.getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE);
+                if (gb != null) {
+                    if (AnnotationUtils.areSameByClass( gb, checkerGuardedByClass) ||
+                        AnnotationUtils.areSameByClass( gb, javaxGuardedByClass) ||
+                        AnnotationUtils.areSameByClass( gb, jcipGuardedByClass)) {
+                        // Do nothing.
+                    } else if (AnnotationUtils.areSameByClass(gb, checkerGuardSatisfiedClass)){
+                        skipCheckPreconditions = true; // Can always dereference if type is @GuardSatisfied
+                    } else {
+                        // Can never dereference for any other types in the @GuardedBy hierarchy
+                        String annotationName = gb.toString();
+                        annotationName = annotationName.substring(annotationName.lastIndexOf('.') + 1 /* +1 to skip the last . as well */);
+                        checker.report(Result.failure(
+                                "cannot.dereference",
+                                tree.toString(),
+                                annotationName), tree);
+                        return;
+                    }
+                }
+
+                if (skipCheckPreconditions == false) {
```

Comment:
What if you changed the name of this boolean and flipped its value?   Like so:
checkPrecondtions = true.
...
else if(...)
checkPrecondtions = false;
...
if(checkPrecondtions)
//check precondtions.